### DISCUSSION TITLE
Question about the usefulness of creating a plugin

### DISCUSSION THREAD

**Discussion by GEJ1** [jspsych/jsPsych#1815] [Ideas]
Hello jsPych comunity,

I'm working in a [Trail Making Test ](https://en.wikipedia.org/wiki/Trail_Making_Test) version using JsPsych. The experiment need to draw filled circles in the screen, the participant should be able to draw in order to connect them and the position of the mouse courser should be tracked. I made all of that using a modified version of the [jspsych-canvas-button-response](https://www.jspsych.org/plugins/jspsych-canvas-button-response/).

My question is the following: It makes sense to made a brand new plugin for Trial Making Test experiments based on the existed canvas plugins? Do you think that will be useful? I think it's a really popular test and the posible variations can be many (e.g.: items, position, background, etc).

Thank you!

Best,

Gustavo.

**Comment by jodeleeuw** [comment]
Sounds great to me @GEJ1 ! I think any plugin that implements a commonly used task in a general way is a great addition.

**Comment by becky-gilbert** [comment]
Hi @GEJ1! I'm not familiar with the Trail Making Test, but I wonder if this is something that could be done using a more general-purpose canvas-drawing plugin, rather than a task-specific plugin? The idea for the canvas-drawing plugin would be to show a canvas that the participant can draw on using their mouse (or finger with touchscreens), and save the final canvas as an image file or [data URI](https://www.tutorialspoint.com/convert-image-to-data-uri-with-javascript#:~:text=To%20convert%20image%20from%20an,the%20toDataURL%20method%20on%20it.). The plugin would include the option to show a stimulus on the canvas for the participant to draw on top of, which could either be an image file or a canvas function (similar to the stimulus functions in the canvas-* plugins). So for your task, I think you would just need to specify the stimulus for each trial as either an image with the filled circles, or a canvas function that draws the circles. Would that work?

The canvas-drawing plugin idea is on our list of thing to add in v7.1, which we won't get to for another few months. But if you'd like to work on this with us then I'm sure we could get it done sooner. I created a canvas-drawing plugin a few years ago for a task where the participant has to draw the hands on a clock face. Unfortunately I wrote this using an older version of jsPsych (v5.0.3 I think?) and the code quality probably isn't great... but I've attached it anyway in case it's useful!

Anyway this is just my two cents. 😃 
[jspsych-canvas-drawing-circle.js.txt](https://github.com/jspsych/jsPsych/files/6468152/jspsych-canvas-drawing-circle.js.txt)

**Reply by madebyafox** [reply]
Hi Becky! Not sure if you're still working on this, but if so... it would definitely be useful in my work studying graph comprehension and production.  In terms of parameters, however, I would need the ability to display a message / instructions / perhaps html elements that are positioned separately from the drawing canvas.  Let me know if you'd like to collaborate!

**Reply by becky-gilbert** [reply]
Hi @madebyafox, we actually added a canvas drawing plugin since my last comment - it's called `sketchpad`: https://www.jspsych.org/7.1/plugins/sketchpad/ The demos at the bottom of the page should give you a quick idea of what the plugin does. Let us know if there's anything else you need!
